server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailConverter.java

```diff
@@ -0,0 +1,170 @@
+    static final AttributeName DELETION_DATE_ATTRIBUTE_NAME = AttributeName.of("deletionDate");
+    static final AttributeName SUBJECT_ATTRIBUTE_VALUE = AttributeName.of("subject");
+
+    private final MailboxId.Factory mailboxIdFactory;
+    private final MessageId.Factory messageIdFactory;
+
+    MailConverter(MailboxId.Factory mailboxIdFactory, MessageId.Factory messageIdFactory) {
+        this.mailboxIdFactory = mailboxIdFactory;
+        this.messageIdFactory = messageIdFactory;
+    }
+
+    MailImpl toMail(DeletedMessage deletedMessage) throws MessagingException {
+        return MailImpl.builder()
+            .name(deletedMessage.getMessageId().serialize())
+            .sender(deletedMessage.getSender())
+            .recipients(deletedMessage.getRecipients())
+            .mimeMessage(new MimeMessage(Session.getDefaultInstance(new Properties()), deletedMessage.getContent().get()))
+            .attribute(OWNER_ATTRIBUTE_NAME.withValue(SerializableUser.toAttributeValue(deletedMessage.getOwner())))
+            .attribute(HAS_ATTACHMENT_ATTRIBUTE_NAME.withValue(AttributeValue.of(deletedMessage.hasAttachment())))
+            .attribute(DELIVERY_DATE_ATTRIBUTE_NAME.withValue(SerializableDate.toAttributeValue(deletedMessage.getDeliveryDate())))
+            .attribute(DELETION_DATE_ATTRIBUTE_NAME.withValue(SerializableDate.toAttributeValue(deletedMessage.getDeletionDate())))
+            .attribute(ORIGIN_MAILBOXES_ATTRIBUTE_NAME.withValue(AttributeValue.of(serializedMailboxIds(deletedMessage))))
+            .attribute(SUBJECT_ATTRIBUTE_VALUE.withValue(subjectAttributeValue(deletedMessage)))
+            .build();
+    }
+
+    DeletedMessage fromMail(Mail mail) {
+        return DeletedMessage.builder()
+            .messageId(messageIdFactory.fromString(mail.getName()))
+            .originMailboxes(retrieveMailboxIds(mail))
+            .user(retrieveOwner(mail))
+            .deliveryDate(retrieveDate(mail, DELIVERY_DATE_ATTRIBUTE_NAME))
+            .deletionDate(retrieveDate(mail, DELETION_DATE_ATTRIBUTE_NAME))
+            .sender(mail.getMaybeSender())
+            .recipients(mail.getRecipients())
+            .content(retrieveContent(mail))
+            .hasAttachment(retrieveHasAttachment(mail))
+            .subject(retrieveSubject(mail))
+            .build();
+    }
+
+    private ImmutableList<AttributeValue<?>> serializedMailboxIds(DeletedMessage deletedMessage) {
+        return deletedMessage.getOriginMailboxes().stream()
+            .map(MailboxId::serialize)
+            .map(AttributeValue::of)
+            .collect(Guavate.toImmutableList());
+    }
+
+    private AttributeValue<Optional<AttributeValue<String>>> subjectAttributeValue(DeletedMessage deletedMessage) {
+        return AttributeValue.of(deletedMessage.getSubject().map(AttributeValue::of));
+    }
+
+    private ContentSupplier retrieveContent(Mail mail) {
```

Comment:
ContentSupplier should in fact be MimeMessageSource, don't you think?

Reply:
Is that a call to be using `MimeMessageWrapper` & related freaks?